modules/uncertainty/R/prep.data.assim.R

```diff
@@ -0,0 +1,82 @@
+##'@title prep.data.assim
+##'@section purpose:
+##'This function sets up a call to PEcAn.assim.sequential::sda.enkf()
+##'
+##'@param settings the PEcAn settings object (a collection of nested lists)
+##'@return None
+##'@export
+##'@author Luke Dramko
+prep.data.assim <- function(settings) {
+  # Obtain real data from the site
+  timestep <- 0.5 # Every half hour = 0.5
+  
+  field_data <- PEcAn.data.atmosphere::download.US_WCr(settings$run$start_date, settings$run$end_date, timestep = timestep)
+  
+  uncertainty_vals <- list()
+  
+  # Creates a proxy row for rbinding
+  sums <- NULL
+  
+  # One vector holds the mean for each variable.
+  obs.mean <- NULL
+  
+  for (i in 1:length(field_data)) {
+    AMF.params <- PEcAn.uncertainty::flux.uncertainty(field_data[[i]], QC = rep(0, length(field_data[[i]])))
+    numvals <- 10;
+    
+    # Create proxy row for rbinding
+    random_mat = NULL
+    new_col = rep(0, length(field_data[[i]]))
+    
+    # Create a new column
+    # i: the particular variable being worked with
+    # j: the column number
+    # k: the row number
+    for (j in 1:numvals) { # number of random numbers
+      obs <- field_data[[i]][!is.na(field_data[[i]])]
+      plot(obs)
+      pos <- obs >= 0
+      
+      res <- obs
+      res[pos]  <- rexp(length(obs[pos]),
+                        1 / (AMF.params$intercept[[1]] + (AMF.params$slopeP[[1]] * obs[pos])))
+      res[!pos]  <- rexp(length(obs[!pos]),
+                         1 / (AMF.params$intercept[[1]] + (AMF.params$slopeN[[1]] * obs[!pos])))
+      
+      random_multiplier <- sample(c(-1,1), length(res), replace = TRUE)
+      simulated <- obs+(random_multiplier*res)
+      
+      sim_idx = 1; # Because NA's are excluded, real time points don't match up exactly.
+      for (k in 1:length(field_data[[i]])) { # k for each real time point
```

Comment:
not following why this step is needed -- the presence of NAs shouldn't change the length of the vector

Reply:
That's because there's a `!is.na() on line 36 (`obs <- field_data[[i]][!is.na(field_data[[i]])]`), which removes the NA's from the vector.  @istfer recommended removing `NA`s before applying the `rexp` functions, perhaps because they're not positive nor negative.  Once the `NA`s are removed, the vector's size is smaller than the original array, and it's no longer a guarantee that the i'th element of the real data corresponds to the i'th element of the simulated data.